learn/appendix_opengl/KHR_debug.md

```diff
@@ -0,0 +1,194 @@
+    DebugType type,             /* Type of the message. */
+    int id,                     /* An ID to associate with this message. */
+    DebugSeverity severity,     /* Severity of the message. */
+    int length,                 /* Length of the message or -1 for automatic. */
+    string message              /* The message to insert. */
+);
+```
+
+### `GL.GetDebugMessageLog()`
+In the abscense of a debug message callback, this function can be used to
+retreive debug messages that have been enqueued by the OpenGL context.
+```cs
+int GL.GetDebugMessageLog(
+    int number,                 /* Number of messages to retreive. */
+    int bufferSize,             /* Message buffer size. */
+    out DebugSource sources,    /* Retreived message sources. */
+    out DebugType types,        /* Retreived message types. */
+    out int ids,                /* Retreived message IDs. */
+    out DebugSeverity severity, /* Retreived message severities */
+    out int length,             /* Retreived message lengths. */
+    out string message          /* Retreived messages. */
+);  /* Returns: The number of messages retreived. */
+
+/* Note: an array overload also exists. */
+```
+
+### Additional Enables
+These additional enables are introduced to the API which allows you to control
+the debug messaging with `GL.Enable()` and `GL.Disable()`.
+
+| EnableCap | Description |
+|:----------|:------------|
+| `EnableCap.DebugOutput` | Enable or disable debug message generation. If disabled, inhibits sending messages to your callback. |
+| `EnableCap.DebugOutputSynchronous` | When enabled, debug messages are sent only to the thread that generated the message. Useful when debugging and viewing the callstack. Hinders performance. |
+
+Object Label API
+----------------
+The object label API allows you to give custom labels to OpenGL objects. These
+labels are then displayed in the debug messages and graphics debuggers.
+Immensely more descriptive than the number assigned by the OpenGL context for
+each OpenGL object or a pointer.
+
+```cs
+/* Label named objects (i.e. objects with handles). */
+void GL.ObjectLabel(
+    ObjectLabelIdentifier identifier,   /* Object family, e.g. a buffer or a texture. */
+    int handle,     /* The object handle. */
+    int length,     /* Length of the label, or -1 for automatic length. */
+    string label    /* The label to give the object. */
+);
+
+/* Label pointer objects */
+void GL.ObjectPtrLabel(
+    IntPtr pointer, /* The pointer to label. */
+    int length,     /* Length of the label, or -1 for automatic length. */
+    string label    /* The label to give to the pointer. */
+);
+```
+
+Debug Group API
+---------------
+The debug group API allows you to group some commands under a label. These
+labels are then displayed by graphics debuggers in a tree fashion. This provides
+additional visual aid and context clues when reading large command lists when
+trying to figure out what went wrong.
```

Comment:
Same here, we want an image to really show how useful this is for a project to have. So we want to show the screenshots of this close to this text.